source/opt/strength_reduction_pass.cpp

```diff
@@ -0,0 +1,211 @@
+}
+
+bool StrengthReductionPass::ReplaceMultiplyByPowerOf2(
+    ir::BasicBlock::iterator* instPtr) {
+  ir::BasicBlock::iterator& inst = *instPtr;
+  assert(inst->opcode() == SpvOp::SpvOpIMul &&
+         "Only works for multiplication of integers.");
+  bool modified = false;
+
+  // Currently only works on 32-bit integers.
+  if (inst->type_id() != int32_type_id_ && inst->type_id() != uint32_type_id_) {
+    return modified;
+  }
+
+  // Check the operands for a constant that is a power of 2.
+  for (int i = 0; i < 2; i++) {
+    uint32_t opId = inst->GetSingleWordInOperand(i);
+    ir::Instruction* opInst = def_use_mgr_->GetDef(opId);
+    if (opInst->opcode() == SpvOp::SpvOpConstant) {
+      // We found a constant operand.
+      uint32_t constVal = opInst->GetSingleWordOperand(2);
+
+      if (IsPowerOf2(constVal)) {
+        modified = true;
+        uint32_t shiftAmount = CountLeadingZeros(constVal);
+        uint32_t shiftConstResultId = GetConstantId(shiftAmount);
+
+        // Create the new instruction.
+        uint32_t newResultId = next_id_++;
+        std::vector<ir::Operand> newOperands;
+        newOperands.push_back(inst->GetInOperand(1 - i));
+        ir::Operand shiftOperand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                                 {shiftConstResultId});
+        newOperands.push_back(shiftOperand);
+        std::unique_ptr<ir::Instruction> newInstruction(
+            new ir::Instruction(SpvOp::SpvOpShiftLeftLogical, inst->type_id(),
+                                newResultId, newOperands));
+
+        // Insert the new instruction and update the data structures.
+        def_use_mgr_->AnalyzeInstDefUse(&*newInstruction);
+        inst = inst.InsertBefore(std::move(newInstruction));
+        ++inst;
+        def_use_mgr_->ReplaceAllUsesWith(inst->result_id(), newResultId);
+
+        // Remove the old instruction.
+        def_use_mgr_->KillInst(&*inst);
+
+        // We do not want to replace the instruction twice if both operands
+        // are constants that are a power of 2.  So we break here.
+        break;
+      }
+    }
+  }
+
+  return modified;
+}
+
+void StrengthReductionPass::FindIntTypesAndConstants() {
+  for (auto typeIter = module_->types_values_begin();
+       typeIter != module_->types_values_end(); ++typeIter) {
+    switch (typeIter->opcode()) {
+      case SpvOp::SpvOpTypeInt:
+        if (typeIter->GetSingleWordOperand(1) == 32) {
+          if (typeIter->GetSingleWordOperand(2) == 1) {
+            int32_type_id_ = typeIter->result_id();
+          } else {
+            uint32_type_id_ = typeIter->result_id();
+          }
+        }
+        break;
+      case SpvOp::SpvOpConstant:
+        if (typeIter->GetSingleWordOperand(0) == uint32_type_id_) {
+          uint32_t value = typeIter->GetSingleWordOperand(2);
+          if (value <= 32)
+            constant_ids_[value] = typeIter->GetSingleWordOperand(1);
```

Comment:
Prefer using result_id() over GetSingleWordOperand(1) as it's a little more readable, and echoes setting of int32_type_id_ and uint32_type_id_